mesonbuild/coredata.py

```diff
@@ -924,6 +924,8 @@ def _parse_section(self, s):
                 raise EnvironmentException(f'Malformed variable name {entry!r} in machine file.')
             # Windows paths...
             value = value.replace('\\', '\\\\')
+            # Fill in path to source root here
+            value = value.replace('@SOURCE_ROOT@', os.getcwd())
```

Comment:
Nothing guarantees that `os.getcwd()` is the source root. You should pass `self.source_dir` from environment.py to `parse_machine_files()` down to MachineFileParser.